File: cadwiki-nuget/cadwiki.AutoCAD2021.Base.Utilities/Shared/XData.vb
Option Strict On
Option Infer Off
Option Explicit On

Imports Autodesk.AutoCAD.ApplicationServices
Imports Autodesk.AutoCAD.DatabaseServices
Imports Autodesk.AutoCAD.EditorInput

Public Class XData
    Public Shared Function GetXDataByRegAppAndKey(doc As Document,
        objId As ObjectId,
        regApp As String,
        xDataKey As String) As XDataFound
        Dim db As Database = doc.Database
        Dim xDataNotFound As New XDataFound
        Using lock As DocumentLock = doc.LockDocument()
            Using t As Transaction = db.TransactionManager.StartTransaction()
                Dim entity As Entity = CType(t.GetObject(objId, OpenMode.ForRead, False), Entity)
                Dim handle As String = entity.Handle.ToString()
                Dim rb As ResultBuffer = entity.XData
                If (rb IsNot Nothing) Then
                    Dim found As XDataFound = GetXDataValueFromBuffer(rb, regApp, xDataKey)
                    Return found
                End If
                t.Commit()
            End Using
        End Using
        Return xDataNotFound
    End Function

    Public Class XDataFound
        Public Index As Integer = -1
        Public Value As String = ""
    End Class

    Public Shared Function GetXDataValueFromBuffer(rb As ResultBuffer, regApp As String, xDataKey As String) _
        As XDataFound
        Dim tvArray As TypedValue() = rb.AsArray
        Dim hasTagBeenFound As Boolean = False
        Dim xDataTag As String = ""
        Dim xDataValue As String = ""
        Dim i As Integer = 0
        Dim found As New XDataFound
        Dim xDataNotFound As New XDataFound
        For Each tv As TypedValue In tvArray
            Dim code As Short = tv.TypeCode
            Select Case code
                Case CType(DxfCode.ExtendedDataRegAppName, Short)
                    Dim appName As String = tv.Value.ToString()
                    If Not appName.Equals(regApp) Then
                        Return xDataNotFound
                    End If
                    ' "{" or "}" are stored as control strings
                Case CType(DxfCode.ExtendedDataControlString, Short)
                    Dim controlString As String = tv.Value.ToString()
                    ' Tags and Values are stored as strings
                Case CType(DxfCode.ExtendedDataAsciiString, Short)
                    If (hasTagBeenFound = False) Then
                        xDataTag = tv.Value.ToString()
                        hasTagBeenFound = True
                    Else
                        xDataValue = tv.Value.ToString()
                        If xDataTag.Equals(xDataKey) Then
                            found.Index = i
                            found.Value = xDataValue
                            Return found
                        End If
                        xDataTag = ""
                        xDataValue = ""
                        hasTagBeenFound = False
                    End If

            End Select
            i = i + 1
        Next
        Return xDataNotFound
    End Function

    Public Shared Sub AddRegAppTableRecord(regAppName As String)
        Dim doc As Document = Application.DocumentManager.MdiActiveDocument

        Dim db As Database = doc.Database

        Using lock As DocumentLock = doc.LockDocument()
            Using t As Transaction = db.TransactionManager.StartTransaction()
                Dim rat As RegAppTable = CType(t.GetObject(db.RegAppTableId, OpenMode.ForRead, False), RegAppTable)
                If Not rat.Has(regAppName) Then
                    rat.UpgradeOpen()
                    Dim ratr As RegAppTableRecord = New RegAppTableRecord()
                    ratr.Name = regAppName
                    rat.Add(ratr)
                    t.AddNewlyCreatedDBObject(ratr, True)
                End If
                t.Commit()
            End Using

        End Using
    End Sub

    Public Shared Function StringValue(appKey As String, dataKey As String, dataValue As String) As ResultBuffer
        Return _
            New ResultBuffer(New TypedValue(1001, appKey),
                New TypedValue(1002, "{"),
                New TypedValue(1000, dataKey),
                New TypedValue(1000, dataValue),
                New TypedValue(1002, "}"))
    End Function

    Public Shared Function CreateAppKey(appKey As String) As ResultBuffer
        Return New ResultBuffer(New TypedValue(1001, appKey))
    End Function


    Public Shared Function CreateStringKeyAndValue(dataKey As String, dataValue As String) As ResultBuffer
        Return _
            New ResultBuffer(New TypedValue(1002, "{"),
                New TypedValue(1000, dataKey),
                New TypedValue(1000, dataValue),
                New TypedValue(1002, "}"))
    End Function

    Public Shared Function CreateTypedValues(dataKey As String, dataValue As String) As TypedValue()
        Dim typesValues() As TypedValue =
            {New TypedValue(1002, "{"), New TypedValue(1000, dataKey), New TypedValue(1000, dataValue),
                New TypedValue(1002, "}")}

        Return typesValues
    End Function

    Public Shared Function AddTypedValues(xDataBuffer As ResultBuffer, typedValues() As TypedValue) As ResultBuffer
        For Each tv As TypedValue In typedValues
            xDataBuffer.Add(tv)
        Next
        Return xDataBuffer
    End Function

    Public Shared Function RemovedTypeValue(xDataBuffer As ResultBuffer, regApp As String, xDataKey As String) _
        As ResultBuffer
        Dim found As XDataFound = GetXDataValueFromBuffer(xDataBuffer, regApp, xDataKey)
        Dim newBuffer As New ResultBuffer()
        If found IsNot Nothing Then
            If Not found.Index = -1 Then
                Dim tvArray As TypedValue() = xDataBuffer.AsArray
                Dim i As Integer = 0
                For Each tv As TypedValue In tvArray
                    If _
                        Not i = found.Index And Not i + 1 = found.Index And Not i - 1 = found.Index And
                            Not i + 2 = found.Index Then
                        newBuffer.Add(tv)
                    End If
                    i = i + 1
                Next
                Return newBuffer
            End If
        End If
        Return xDataBuffer
    End Function

    Public Shared Function GetSsOfMatchingXdata(doc As Document,
        ss As SelectionSet,
        xDataRegAppKey As String,
        xDataKey As String,
        xdataValue As String) As List(Of ObjectId)
        Try
            Dim matchingXDataObjectIdList As New List(Of ObjectId)
            Dim db As Database = doc.Database
            Using lock As DocumentLock = doc.LockDocument()
                Using t As Transaction = db.TransactionManager.StartTransaction()
                    Dim currentSpace As BlockTableRecord = CType(t.GetObject(db.CurrentSpaceId, OpenMode.ForWrite),
                        BlockTableRecord)
                    For Each objId As ObjectId In ss.GetObjectIds
                        Dim entity As Entity = CType(t.GetObject(objId, OpenMode.ForWrite), Entity)
                        Dim existingXData As ResultBuffer = entity.XData
                        Dim xdataFound As XDataFound = GetXDataValueFromBuffer(existingXData, xDataRegAppKey, xDataKey)
                        If xdataFound IsNot Nothing Then
                            If Not xdataFound.Index = -1 Then
                                If xdataFound.Value = xdataValue Then
                                    matchingXDataObjectIdList.Add(entity.Id)
                                End If
                            End If
                        End If
                    Next
                    t.Commit()
                End Using
            End Using
            Return matchingXDataObjectIdList
        Catch ex As Exception
            Throw New Exception("GetMatchingXdata error: " + ex.Message)
        End Try
        Return Nothing
    End Function

    Public Shared Function SetXdataValue(doc As Document,
        existingBuffer As ResultBuffer,
        xDataRegAppKey As String,
        xDataKey As String,
        xdataValue As String) As ResultBuffer
        Dim existingDataEntry As String =
            XData.GetXDataValueFromBuffer(existingBuffer, xDataRegAppKey, xDataKey).Value
        Dim newKeyAndValue() As TypedValue = XData.CreateTypedValues(xDataKey, xdataValue)
        Dim newBuffer As ResultBuffer = existingBuffer
        If existingDataEntry Is Nothing Then
            newBuffer = XData.AddTypedValues(newBuffer, newKeyAndValue)
        Else
            newBuffer = XData.RemovedTypeValue(newBuffer, xDataRegAppKey, xDataKey)
            newBuffer = XData.AddTypedValues(newBuffer, newKeyAndValue)
        End If
        Return newBuffer
    End Function
End Class

File: cadwiki-nuget/cadwiki.AC/Shared/XData.cs
using System;
using System.Collections.Generic;
using Autodesk.AutoCAD.ApplicationServices;
using Autodesk.AutoCAD.DatabaseServices;
using Autodesk.AutoCAD.EditorInput;

namespace cadwiki.AC
{

    public class XData
    {
        public static XDataFound GetXDataByRegAppAndKey(Document doc, ObjectId objId, string regApp, string xDataKey)
        {
            var db = doc.Database;
            var xDataNotFound = new XDataFound();
            using (var @lock = doc.LockDocument())
            {
                using (var t = db.TransactionManager.StartTransaction())
                {
                    Entity entity = (Entity)t.GetObject(objId, global::Autodesk.AutoCAD.DatabaseServices.OpenMode.ForRead, false);
                    string handle = entity.Handle.ToString();
                    var rb = entity.XData;
                    if (rb is not null)
                    {
                        var found = GetXDataValueFromBuffer(rb, regApp, xDataKey);
                        return found;
                    }
                    t.Commit();
                }
            }
            return xDataNotFound;
        }

        public class XDataFound
        {
            public int Index = -1;
            public string Value = "";
        }

        public static XDataFound GetXDataValueFromBuffer(ResultBuffer rb, string regApp, string xDataKey)
        {
            TypedValue[] tvArray = rb.AsArray();
            bool hasTagBeenFound = false;
            string xDataTag = "";
            string xDataValue = "";
            int i = 0;
            var found = new XDataFound();
            var xDataNotFound = new XDataFound();
            foreach (TypedValue tv in tvArray)
            {
                short code = tv.TypeCode;
                switch (code)
                {
                    case 1001:
                        {
                            string appName = tv.Value.ToString();
                            if (!appName.Equals(regApp))
                            {
                                return xDataNotFound;
                                // "{" or "}" are stored as control strings
                            }

                            break;
                        }

                    case 1002:
                        {
                            // Tags and Values are stored as strings
                            string controlString = tv.Value.ToString();
                            break;
                        }

                    case 1000:
                        {
                            if (hasTagBeenFound == false)
                            {
                                xDataTag = tv.Value.ToString();
                                hasTagBeenFound = true;
                            }
                            else
                            {
                                xDataValue = tv.Value.ToString();
                                if (xDataTag.Equals(xDataKey))
                                {
                                    found.Index = i;
                                    found.Value = xDataValue;
                                    return found;
                                }
                                xDataTag = "";
                                xDataValue = "";
                                hasTagBeenFound = false;
                            }

                            break;
                        }

                }
                i = i + 1;
            }
            return xDataNotFound;
        }

        public static void AddRegAppTableRecord(string regAppName)
        {
            var doc = global::Autodesk.AutoCAD.ApplicationServices.Core.Application.DocumentManager.MdiActiveDocument;

            var db = doc.Database;

            using (var @lock = doc.LockDocument())
            {
                using (var t = db.TransactionManager.StartTransaction())
                {
                    RegAppTable rat = (RegAppTable)t.GetObject(db.RegAppTableId, global::Autodesk.AutoCAD.DatabaseServices.OpenMode.ForRead, false);
                    if (!rat.Has(regAppName))
                    {
                        rat.UpgradeOpen();
                        var ratr = new RegAppTableRecord();
                        ratr.Name = regAppName;
                        rat.Add(ratr);
                        t.AddNewlyCreatedDBObject(ratr, true);
                    }
                    t.Commit();
                }

            }
        }

        public static ResultBuffer StringValue(string appKey, string dataKey, string dataValue)
        {
            return new ResultBuffer(new TypedValue(1001, appKey), new TypedValue(1002, "{"), new TypedValue(1000, dataKey), new TypedValue(1000, dataValue), new TypedValue(1002, "}"));
        }

        public static ResultBuffer CreateAppKey(string appKey)
        {
            return new ResultBuffer(new TypedValue(1001, appKey));
        }


        public static ResultBuffer CreateStringKeyAndValue(string dataKey, string dataValue)
        {
            return new ResultBuffer(new TypedValue(1002, "{"), new TypedValue(1000, dataKey), new TypedValue(1000, dataValue), new TypedValue(1002, "}"));
        }

        public static TypedValue[] CreateTypedValues(string dataKey, string dataValue)
        {
            TypedValue[] typesValues = new TypedValue[] { new TypedValue(1002, "{"), new TypedValue(1000, dataKey), new TypedValue(1000, dataValue), new TypedValue(1002, "}") };

            return typesValues;
        }

        public static ResultBuffer AddTypedValues(ResultBuffer xDataBuffer, TypedValue[] typedValues)
        {
            foreach (TypedValue tv in typedValues)
                xDataBuffer.Add(tv);
            return xDataBuffer;
        }

        public static ResultBuffer RemovedTypeValue(ResultBuffer xDataBuffer, string regApp, string xDataKey)
        {
            var found = GetXDataValueFromBuffer(xDataBuffer, regApp, xDataKey);
            var newBuffer = new ResultBuffer();
            if (found is not null)
            {
                if (!(found.Index == -1))
                {
                    TypedValue[] tvArray = xDataBuffer.AsArray();
                    int i = 0;
                    foreach (TypedValue tv in tvArray)
                    {
                        if (!(i == found.Index) & !(i + 1 == found.Index) & !(i - 1 == found.Index) & !(i + 2 == found.Index))
                        {
                            newBuffer.Add(tv);
                        }
                        i = i + 1;
                    }
                    return newBuffer;
                }
            }
            return xDataBuffer;
        }

        public static List<ObjectId> GetSsOfMatchingXdata(Document doc, SelectionSet ss, string xDataRegAppKey, string xDataKey, string xdataValue)
        {
            var matchingXDataObjectIdList = new List<ObjectId>();
            var db = doc.Database;
            using (var @lock = doc.LockDocument())
            {
                using (var t = db.TransactionManager.StartTransaction())
                {
                    BlockTableRecord currentSpace = (BlockTableRecord)t.GetObject(db.CurrentSpaceId, global::Autodesk.AutoCAD.DatabaseServices.OpenMode.ForWrite);
                    foreach (ObjectId objId in ss.GetObjectIds())
                    {
                        Entity entity = (Entity)t.GetObject(objId, global::Autodesk.AutoCAD.DatabaseServices.OpenMode.ForWrite);
                        var existingXData = entity.XData;
                        if (existingXData == null)
                        {
                            continue;
                        }
                        var xdataFound = GetXDataValueFromBuffer(existingXData, xDataRegAppKey, xDataKey);
                        if (xdataFound is not null)
                        {
                            if (!(xdataFound.Index == -1))
                            {
                                if ((xdataFound.Value ?? "") == (xdataValue ?? ""))
                                {
                                    matchingXDataObjectIdList.Add(entity.Id);
                                }
                            }
                        }
                    }
                    t.Commit();
                }
            }
            return matchingXDataObjectIdList;
        }

        public static ResultBuffer SetXdataValue(Document doc, ResultBuffer existingBuffer, string xDataRegAppKey, string xDataKey, string xdataValue)
        {
            string existingDataEntry = GetXDataValueFromBuffer(existingBuffer, xDataRegAppKey, xDataKey).Value;
            TypedValue[] newKeyAndValue = CreateTypedValues(xDataKey, xdataValue);
            var newBuffer = existingBuffer;
            if (existingDataEntry is null)
            {
                newBuffer = AddTypedValues(newBuffer, newKeyAndValue);
            }
            else
            {
                newBuffer = RemovedTypeValue(newBuffer, xDataRegAppKey, xDataKey);
                newBuffer = AddTypedValues(newBuffer, newKeyAndValue);
            }
            return newBuffer;
        }
    }
}